go/tools/go-upgrade/go_upgrade.go

```diff
@@ -0,0 +1,453 @@
+	if err != nil {
+		return nil, err
+	}
+
+	var latestGoReleases []latestGolangRelease
+	err = json.Unmarshal(body, &latestGoReleases)
+	if err != nil {
+		return nil, err
+	}
+
+	var versions version.Collection
+	for _, release := range latestGoReleases {
+		if !release.Stable {
+			continue
+		}
+		if !strings.HasPrefix(release.Version, "go") {
+			return nil, fmt.Errorf("golang version malformatted: %s", release.Version)
+		}
+		newVersion, err := version.NewVersion(release.Version[2:])
+		if err != nil {
+			return nil, err
+		}
+		versions = append(versions, newVersion)
+	}
+	return versions, nil
+}
+
+// chooseNewVersion decides what will be the next version we're going to use in our codebase.
+// Given the current Golang version, the available latest versions and whether we allow major upgrade or not,
+// chooseNewVersion will return either the new version or nil if we cannot/don't need to upgrade.
+func chooseNewVersion(curVersion *version.Version, latestVersions version.Collection, allowMajorUpgrade bool) *version.Version {
+	selectedVersion := curVersion
+	for _, latestVersion := range latestVersions {
+		if !allowMajorUpgrade && !isSameMajorVersion(latestVersion, selectedVersion) {
+			continue
+		}
+		if latestVersion.GreaterThan(selectedVersion) {
+			selectedVersion = latestVersion
+		}
+	}
+	// No change detected, return nil meaning that we do not want to have a new Golang version.
+	if selectedVersion.Equal(curVersion) {
+		return nil
+	}
+	return selectedVersion
+}
+
+// replaceGoVersionInCodebase goes through all the files in the codebase where the
+// Golang version must be updated
+func replaceGoVersionInCodebase(old, new *version.Version, noWorkflowUpdate bool) error {
+	explore := []string{
+		"./test/templates",
+		"./build.env",
+		"./docker/bootstrap/Dockerfile.common",
+	}
+	if !noWorkflowUpdate {
+		explore = append(explore, "./.github/workflows")
+	}
+	filesToChange, err := getListOfFilesInPaths(explore)
+	if err != nil {
+		return err
+	}
+
+	for _, fileToChange := range filesToChange {
+		err = replaceInFile(
+			[]string{old.String()},
+			[]string{new.String()},
+			fileToChange,
+		)
+		if err != nil {
+			return err
+		}
+	}
+
+	if !isSameMajorVersion(old, new) {
+		err = replaceInFile(
+			[]string{fmt.Sprintf("%d.%d", old.Segments()[0], old.Segments()[1])},
+			[]string{fmt.Sprintf("%d.%d", new.Segments()[0], new.Segments()[1])},
+			"./go.mod",
+		)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func updateBootstrapVersionInCodebase(old, new float64, newGoVersion *version.Version) error {
+	files, err := getListOfFilesInPaths([]string{
+		"./docker/base",
+		"./docker/lite",
+		"./docker/local",
+		"./docker/vttestserver",
+		"./Makefile",
+		"./test/templates",
+	})
+	if err != nil {
+		return err
+	}
+
+	const btv = "bootstrap_version="
+	oldReplace := fmt.Sprintf("%s%-1g", btv, old)
+	newReplace := fmt.Sprintf("%s%-1g", btv, new)
+	for _, file := range files {
```

Comment:
The Makefile at least has this: https://github.com/vitessio/vitess/blob/main/Makefile#L272-L306

Shouldn't we be replacing the `BOOTSTRAP_VERSION` variable? And another note here about brittleness as we should at least allow whitespace.

Reply:
The code below replaces the `BOOTSTRAP_VERSION` variable in the `Makefile`. `replaceInFile` looks for both `BOOTSTRAP_VERSION=` and `bootstrap_version=`.

I'll update this to use regexp though.